platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,927 @@
+- Render the form navigation buttons "Back" and "Continue"
+- Prevent navigation if there are validation errors
+
+**A note on cloning children**
+To pass the form state to `Page`'s children, it will clone all its children and
+pass the form state props along with the child's given props. Implementation
+will look something like the following:
+```js
+const newChildren = React.Children.map(
+  this.props.children,
+  child => React.cloneElement(child, { formData, pathParams }));
+```
+
+This means developers can create form pages in a couple different ways. 
+
+_Using the form state:_
+```jsx
+<Page>
+  {({ formData, pathParams }) => (/* Use the form state to render the correct fields */)}
+</Page
+```
+
+_Not using the form state:_
+```jsx
+<Page>
+  <StringField name="foo" />
+  <StringField name="bar" />
+</Page
+```
+
+In the above example, both `StringField`s will have the following props:
+- `name`
+- `formData`
+- `pathParams`
+
+The form state props `formData` and `pathParams` will simply go unused.
+
+##### `Route` Component
+**Purpose:**
+The `Route` will be re-exported from `react-router-dom`. The purpose of
+re-exporting is twofold:
+- Simplifies imports
+- Ensures only one version of `react-router-dom` is used
+
+**Usage requirements:**
+- `Route` may only be a child of `Router`
+
+**Why not just use `Page`?**
+The `Chapter` and `Page` components are meant to be used to define pages where
+the user may enter data. These pages are sequential and have back and continue
+buttons. `Route`s are used for pages in the application that live outside that
+form flow. These pages may include:
+- `/introduction`
+- `/form-saved`
+- `/error`
+
+##### Functionality: Array pages
+Array pages are one or more pages that are looped through in a form flow, once
+per item in the array. To help understand the nuances of array pages, consider
+the following example:
+
+```jsx
+<Page
+  path="/marriages/:marriageIndex/foo"
+  pathParams={{
+    marriageIndex: 'marriages'
+  }}>
+  {({ formData, pathParams }) => (
+    <FieldArray
+      name="marriages"
+      render={arrayHelpers => (
+        <Field name={`marriages.${pathParams.marriageIndex}.name.first`} />
+      )}
+    />
+  )}
+</Page>
+```
+
+For array pages to work we need a couple core things:
+- The path for the page
+- The data for the current item in the array
+
+To get these pieces, `Page` will accept the `path` and `pathParams` props.
+
+`path` will be the URL within the application and must (for array pages) contain
+URL parameters (term pending). URL parameters must be the only token between
+`/`s or at the end of the URL. For example, `/:index/foo` and `/:index` are
+valid path parameters. `/my-:index-item` is an invalid path parameter.
+
+`pathParams` will be an object whose keys map to the path parameters and whose
+values map to the path to the form data. In the above example, the `pathParams`
+will assign `:marriageIndex` in the URL a number based in the current item in
```

Comment:
can the index be named anything? Is the only requirement the `:` at the beginning? Or does the word "index" have to be included?

Reply:
Yeah, whatever you want to name it. The only requirement is that the URL segment start with a `:`. I'll clarify in the document. :+1: